cairo-verifier/src/utils.rs

```diff
@@ -0,0 +1,223 @@
+    if bytes.len() < proof_len {
+        return Err(VerifierError(format!("Error reading proof from file: {}", proof_path)));
+    }
+    
+    let Ok((proof, _)) = bincode::serde::decode_from_slice(
+        &bytes[0..proof_len],
+        bincode::config::standard(),
+    ) else {
+        return Err(VerifierError(format!("Error reading proof from file: {}", proof_path)));
+    };
+    
+    // PublicInputs len was stored as an u32, 4u8 needs to be read
+    let pub_inputs_len = u32::from_le_bytes(bytes[proof_len..proof_len+4].try_into().unwrap()) as usize;
+    let pub_inputs_bytes = &bytes[proof_len+4..proof_len+4+pub_inputs_len];
+    
+    let Ok((pub_inputs, _)) =
+    bincode::serde::decode_from_slice(pub_inputs_bytes, bincode::config::standard())
+    else {
+        return Err(VerifierError(format!("Error reading proof from file: {}", proof_path)));
+    };
+    let program_output_bytes = &bytes[proof_len+4+pub_inputs_len..];
+
+    let Ok((program_output, _)) =
+        bincode::serde::decode_from_slice::<HyleOutput<Event>, _>(program_output_bytes, bincode::config::standard())
+    else {
+        return Err(VerifierError(format!("Error reading proof from file: {}", proof_path)));
+    };
+
+    if verify_cairo_proof(&proof, &pub_inputs, &proof_options){
+        return Ok(serde_json::to_string(&program_output)?);
+    }
+    else {
+        return Err(VerifierError(format!("Error reading proof from file: {}", proof_path)));
+    }
+}
+
+
+pub fn prove(trace_data: Vec<u8>, memory_data: Vec<u8>, output: &str) -> Result<Vec<u8>, VerifierError> {
+    let proof_options = ProofOptions::new_secure(SecurityLevel::Conjecturable100Bits, 3);
+    let Some((proof, pub_inputs)) = generate_proof_from_trace(
+        &trace_data,
+        &memory_data,
+        &proof_options,
+    ) else {
+        return Err(VerifierError("Error generation prover args".to_string()));
+    };
+    let proof = write_proof(proof, pub_inputs, output);
+    Ok(proof)
+}
+
+pub fn generate_proof_from_trace(
+    trace_data: &Vec<u8>,
+    memory_data: &Vec<u8>,
+    proof_options: &ProofOptions,
+) -> Option<(
+    StarkProof<Stark252PrimeField, Stark252PrimeField>,
+    PublicInputs,
+)> {
+    // ## Generating the prover args
+    let register_states = RegisterStates::from_bytes_le(trace_data).expect("Cairo trace data incorrect");
+    let memory = CairoMemory::from_bytes_le(memory_data).expect("Cairo memory data incorrect");
+
+    // data length
+    let data_len = 0_usize;
+    let mut pub_inputs = PublicInputs::from_regs_and_mem(&register_states, &memory, data_len);
+
+
+    let main_trace = build_main_trace(&register_states, &memory, &mut pub_inputs);
+
+
+    // ## Prove
+    let proof = match generate_cairo_proof(&main_trace, &pub_inputs, proof_options) {
+        Ok(p) => p,
+        Err(err) => {
+            eprintln!("Error generating proof: {:?}", err);
+            return None;
+        }
+    };
+
+    Some((proof, pub_inputs))
+}
+
+fn write_proof(
```

Comment:
So the goal of this function is to concatenate everything necessary to the proof verification right ?
proof(pub_inputs) -> output

Reply:
verify proof then decodes what this function has stored into a file